### DISCUSSION TITLE
DataGrid Predictive Automatic Column Width

### DISCUSSION THREAD

**Discussion by nixtar** [AvaloniaUI/Avalonia#17397] [Q&A]
Hello,

I have an app with a DataGrid with a lot of rows and the columns have inconsistent data from a width perspective.

I'd like the DataDrid Columns to behave as though the width is set to Auto for all the data from the start.
As I understand it this might be difficult due to virtualisation preventing the width from calculating until you scroll.

WPF has the same issue and in the past with WPF, I've implemented a hack where I jump over all the columns and then over all the values and calculate the max width by creating a TextBlock with the content and measuring etc.
I'm hoping there is a nicer way?

I've written a quick and dirty reproduction in case anyone wants to have a play.

Avalonia 11.2.0 MVVM Template using CommunityToolkit.

MainWindowViewModel.cs
```c#
using System;
using System.Collections.ObjectModel;
using System.Text;
using CommunityToolkit.Mvvm.ComponentModel;

namespace Avalonia.DataGrid.ViewModels;

public partial class MainWindowViewModel : ViewModelBase
{
    public MainWindowViewModel()
    {
        GreetingCollection = new ObservableCollection<string>();
        GenerateGreetings(100);
    }

    [ObservableProperty] private ObservableCollection<string> _greetingCollection;
    
    
    public void GenerateGreetings(int numberOfGreetings)
    {
        Random random = new Random();
        string[] greetingBases = { "Hello", "Hi", "Hey", "Greetings", "Salutations", "Howdy", "Welcome" };

        for (int i = 0; i < numberOfGreetings; i++)
        {
            // Choose a random greeting base
            string baseGreeting = greetingBases[random.Next(greetingBases.Length)];

            // Create a greeting with increasing length
            StringBuilder greetingBuilder = new StringBuilder(baseGreeting);
            int repeatCount = i / 10; // Increase length every 10 greetings

            for (int j = 0; j < repeatCount; j++)
            {
                greetingBuilder.Append(", ").Append(baseGreeting);
            }

            GreetingCollection.Add(greetingBuilder.ToString());
        }
    }
}
```

MainWindow.axaml
```c#
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:Avalonia.DataGrid.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="Avalonia.DataGrid.Views.MainWindow"
        x:DataType="vm:MainWindowViewModel"
        Icon="/Assets/avalonia-logo.ico"
        Title="Avalonia.DataGrid">

    <Design.DataContext>
        <vm:MainWindowViewModel/>
    </Design.DataContext>

    <DataGrid Margin="20" ItemsSource="{Binding GreetingCollection}" ColumnWidth="Auto">
        <DataGrid.Columns>
            <DataGridTextColumn Header="Greeting" Binding="{Binding}"/>
        </DataGrid.Columns>
    </DataGrid>

</Window>
```

App.axaml
```c#
<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             x:Class="Avalonia.DataGrid.App"
             RequestedThemeVariant="Default">
             <!-- "Default" ThemeVariant follows system theme variant. "Dark" or "Light" are other available options. -->
  
    <Application.Styles>
        <FluentTheme />
        <StyleInclude Source="avares://Avalonia.Controls.DataGrid/Themes/Fluent.xaml"/>
    </Application.Styles>
</Application>
```

Video of the behaviour (See the column width grow as I scroll)

https://github.com/user-attachments/assets/ebe3f980-1eff-4af7-bfa4-ac1c61936812

**Comment by Coloryr** [comment]
Do you mean it can automatically shrink?

**Reply by nixtar** [reply]
No I'd like it to start with the maximum size without having to scroll through the all the data.

**Reply by nixtar** [reply]
Just like if vitrualisation was disabled.

**Reply by Coloryr** [reply]
At the beginning, the textblock control will not set the longest string, so it will not automatically to the maximum. To do this, it's necessary to `foreach` all strings every time when the source change, which may cause performance issues.

**Reply by nixtar** [reply]
Yeah I noted this as a possible solution that I've done before with WPF. Was hoping there might be something more elegant.

I wonder if it's possible to disable vitrualisation before loading the data then enable it after. (not sure if avalonia even supports doing this)

Won't be great for data sources that change a lot, but my use case in particular is static data after the initial load.

**Comment by timunie** [comment]
Maybe formattedtext is enough to calculate the expected width but needs to be tested
